### PR TITLE
Examples: Box Projected CubeMap Environment Mapping

### PR DESCRIPTION
**Pull request by sneha-belkhale**
I made an example demo of Box Projected CubeMap Environment Mapping, as a follow up to the discussion in #15704. 

The scene uses a `CubeCamera` to generate an environment map - and applies the environment map to a ground mesh that has a `MeshPhysicalMaterial` with custom shader injections that implement Box Projection, following this article: 
http://devlog-martinsh.blogspot.com/2011/09/box-projected-cube-environment-mapping.html

There is also a toggle control to contrast what the environment map would look like without Box Projection. 

screenshot of the scene: 

screen shot 2019-03-04 at 12 48 45 am

Let me know if the shader injection snippets should be moved to another file -- although I am not sure where it should go .. maybe in the `examples/js/shaders` directory?

### CODE REVIEW

**Comment by Mugen87** [comment]
https://raw.githack.com/sneha-belkhale/three.js/bpcem/examples/webgl_box_projected_cubemap_environment_mapping.html

File: examples/webgl_box_projected_cubemap_environment_mapping.html [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+
+				var defaultMat = new THREE.MeshPhysicalMaterial({
+					roughness: 1,
+					envMap: cubeCamera.renderTarget.texture,
+					roughnessMap: rMap,
+				});
+
+				var boxProjectedMat = new THREE.MeshPhysicalMaterial({
+					color: new THREE.Color('#ffffff'),
+			    roughness: 1,
```
Comment (Mugen87) [inline examples/webgl_box_projected_cubemap_environment_mapping.html]: There are some minor formatting issues in your file. Please ensure to use only hard tabs.

File: examples/webgl_box_projected_cubemap_environment_mapping.html [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+					shader.fragmentShader = shader.fragmentShader.replace(
+						'#include <envmap_physical_pars_fragment>',
+						envmapPhysicalParsReplace
+					);
+
+					materialShader = shader;
+
+				};
+
+			  groundPlane = new THREE.Mesh( new THREE.PlaneGeometry( 200, 100, 100 ), boxProjectedMat );
```
Comment (Mugen87) [inline examples/webgl_box_projected_cubemap_environment_mapping.html]: Please use `PlaneBufferGeometry` here.

File: examples/webgl_box_projected_cubemap_environment_mapping.html [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+				groundPlane.rotateX( - Math.PI / 2 );
+				groundPlane.position.set( 0, - 49, 0 );
+				groundPlane.geometry.computeVertexNormals();
```
Comment (Mugen87) [inline examples/webgl_box_projected_cubemap_environment_mapping.html]: Why are you computing the vertex normals here?

File: examples/webgl_box_projected_cubemap_environment_mapping.html [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+				cameraControls.target.set( 0, -10, 0 );
+				cameraControls.maxDistance = 400;
+				cameraControls.minDistance = 10;
+				cameraControls.update();
+
+				// cube camera for environment map
+				cubeCamera = new THREE.CubeCamera( 1, 1000, 512 );
+				cubeCamera.renderTarget.texture.generateMipmaps = true;
+				cubeCamera.renderTarget.texture.minFilter = THREE.LinearMipMapLinearFilter;
+				cubeCamera.renderTarget.texture.magFilter = THREE.LinearFilter;
```
Comment (Mugen87) [inline examples/webgl_box_projected_cubemap_environment_mapping.html]: `THREE.LinearFilter` should be already the default value. No need to set it then.

File: examples/webgl_box_projected_cubemap_environment_mapping.html (hunk 3 above)
Comment (sneha-belkhale) [reply examples/webgl_box_projected_cubemap_environment_mapping.html]: My bad, forgot to remove this line, I think it's remnants from when I was using another geo.

**Review by Mugen87 — APPROVED** [review]

**Comment by Mugen87** [comment]
Oh, I've noticed on more thing: I would rename the example to `webgl_box_projected_cubemap` in order to avoid the line breakage here:

image

**Comment by sneha-belkhale** [comment]
@Mugen87 yes agreed, thanks for reviewing:)

File: examples/webgl_materials_envmaps_bpcem.html [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+			var FAR = 800;
+
+			var camera, scene, renderer;
+
+			var cameraControls;
+
+			var groundPlane, smallSphere, wallMat;
+
+			init();
+			animate();
```
Comment (WestLangley) [inline examples/webgl_materials_envmaps_bpcem.html]: You shouldn't need an animation loop for a static scene. Instead, do this:
```js
controls.addEventListener( 'change', render );
```

**Comment by WestLangley** [comment]
How about
`webgl_materials_envmaps_bpcem` -> `webgl_materials_envmaps_box_projected`

**Comment by Mugen87** [comment]
If it fits in one line I'm okay with that^^

**Comment by WestLangley** [comment]
>If it fits in one line I'm okay with that^^

`webgl_materials_envmaps_parallax` is shorter.

File: examples/webgl_materials_envmaps_bpcem.html (hunk 5 above)
Comment (sneha-belkhale) [reply examples/webgl_materials_envmaps_bpcem.html]: oh nice, never thought of it!

**Comment by sneha-belkhale** [comment]
`webgl_materials_envmaps_box_projected` didn't fit on one line, so I went with `webgl_materials_envmaps_parallax` ~

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**